scripts/convert.py

```diff
@@ -0,0 +1,244 @@
+# Standard library
+from os import path, walk, remove, makedirs
+
+# Third-party
+from astropy import log as logger
+logger.setLevel('INFO')
+
+from nbconvert.preprocessors import ExecutePreprocessor, CellExecutionError
+from nbconvert.exporters import RSTExporter
+from nbconvert.writers import FilesWriter
+import nbformat
+
+IPYTHON_VERSION = 4
+
+class NBConverter(object):
+
+    def __init__(self, nb_path, output_path=None, template_file=None,
+                 overwrite=False):
+        self.nb_path = path.abspath(nb_path)
+        fn = path.basename(self.nb_path)
+        self.path_only = path.dirname(self.nb_path)
+        self.nb_name, _ = path.splitext(fn)
+
+        if output_path is not None:
+            self.output_path = output_path
+            makedirs(self.output_path, exist_ok=True)
+        else:
+            self.output_path = self.path_only
+
+        if template_file is not None:
+            self.template_file = path.abspath(template_file)
+        else:
+            self.template_file = None
+
+        self.overwrite = overwrite
+
+        # the executed notebook
+        self._executed_nb_path = path.join(self.output_path,
+                                           'exec_{0}'.format(fn))
+
+        logger.info('Processing notebook {0} (in {1})'.format(fn,
+                                                              self.path_only))
+
+    def execute(self, write=True):
+        """
+        Execute the specified notebook file, and optionally write out the
+        executed notebook to a new file.
+
+        Parameters
+        ----------
+        write : bool, optional
+            Write the executed notebook to a new file, or not.
+
+        Returns
+        -------
+        executed_nb_path : str, ``None``
+            The path to the executed notebook path, or ``None`` if
+            ``write=False``.
+
+        """
+
+        if path.exists(self._executed_nb_path) and not self.overwrite:
+            logger.debug("Executed notebook already exists at {0}. Use "
+                         "overwrite=True or --overwrite (at cmd line) to re-run"
+                         .format(self._executed_nb_path))
+            return self._executed_nb_path
+
+        # Execute the notebook
+        logger.debug('Executing notebook...')
+        executor = ExecutePreprocessor(timeout=900, kernel_name='python3')
+        with open(self.nb_path) as f:
+            nb = nbformat.read(f, as_version=IPYTHON_VERSION)
+
+        try:
+            executor.preprocess(nb, {'metadata': {'path': self.path_only}})
+        except CellExecutionError:
+            # TODO: should we fail fast and raies, or record all errors?
```

Comment:
I think "record all errors" is better, but that can await a later PR if it seems too much work to manage right now.

Reply:
👍